Defensive Programming in Substrate Reference Document

CrackTheCode016 opened the pull request:
_This PR is being continued from https://github.com/paritytech/polkadot-sdk/pull/2206, which was closed when the developer_hub was merged._
closes https://github.com/paritytech/polkadot-sdk-docs/issues/44

---
# Description

This PR adds a reference document to the `developer-hub` crate (see https://github.com/paritytech/polkadot-sdk/pull/2102).  This specific reference document covers defensive programming practices common within the context of developing a runtime with Substrate.

In particular, this covers the following areas: 

- Default behavior of how Rust deals with numbers in general
- How to deal with floating point numbers in runtime / fixed point arithmetic
- How to deal with Integer overflows
- General "safe math" / defensive programming practices for common pallet development scenarios
- Defensive traits that exist within Substrate, i.e., `defensive_saturating_add `, `defensive_unwrap_or`
- More general defensive programming examples (keep it concise)
- Link to relevant examples where these practices are actually in production / being used
- Unwrapping (or rather lack thereof) 101

todo
-- 
- [x] Apply feedback from previous PR
- [x] This may warrant a PR to append some of these docs to `sp_arithmetic`

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4603288

ggwpez commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+//! As our runtime should _never_ panic; this includes carefully handling [`Result`]/[`Option`]
```
License is missing.

ggwpez commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+//! As our runtime should _never_ panic; this includes carefully handling [`Result`]/[`Option`]
+//! types, eliminating the possibility of integer overflows, converting between number types, or
+//! even handling floating point usage with fixed point arithmetic to mitigate issues that come with
```
```suggestion
//! even replacing floating point usage with fixed point arithmetic to mitigate issues that come with
```

ggwpez commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+//! floating point calculations.
+//!
+//! Intentional and predictable design should be our first and foremost
+//! property for ensuring a well running, safely designed system.
```
We normally try to do line breaks at exactly 100 chars in the codebase.

kianenigma replied:
https://marketplace.visualstudio.com/items?itemName=stkb.rewrap will do the work.

CrackTheCode016 replied:
Forgot to fmt, my bad, and that extension is super useful!

ggwpez commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+//! When developing within the context of the a runtime, there is one golden rule:
+//!
+//! ***DO NOT PANIC,*** most of the time. There are some exceptions, such as critical operations being actually more
+//! dangerous than allowing the node to continue running (block authoring, consensus, etc).
```
The node is outside of the runtime and does not matter as much when it panics. The node is also unaffected when the runtime panics, but it can open a DoS vector and effectively allow an attacker to stall the chain.

ggwpez commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+//!
+//! - Avoid panickable operations, such as directly using `unwrap()` for a [`Result`], common errors such as accessing collections out of bounds (using safer methods to access collection types, i.e., `get()`).
+//! - It may be acceptable to use `except()`, but only if one is completely certain (and has
+//!   performed a check beforehand) that a value won't panic upon unwrapping.
```
Even then; one developer may be certain of it, but in a future iteration of changes someone else misses the assumption and changes it in an unsafe way.  
I think in general it should be avoided as much as possible - even when one is certain.

ggwpez commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+//! - Avoid panickable operations, such as directly using `unwrap()` for a [`Result`], common errors such as accessing collections out of bounds (using safer methods to access collection types, i.e., `get()`).
+//! - It may be acceptable to use `except()`, but only if one is completely certain (and has
+//!   performed a check beforehand) that a value won't panic upon unwrapping.
+//! - If you are writing a function that could panic, [be sure to document it!](https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html#documenting-components)
```
I think we should not write such functions in general.  
However, there are some. And normally we suffix them with `_unchecked` to make it obvious, since most devs dont read the docs 😆

ggwpez commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+//!
+//! The [`Defensive`](frame::traits::Defensive) trait and its companions,
+//! [`DefensiveOption`](frame::traits::DefensiveOption),
+//! [`DefensiveResult`](frame::traits::DefensiveResult) can be used to defensively unwrap
```
I dont know if we have them listed somewhere, but there are a few more: ![Screenshot 2024-01-02 at 11 38 07](https://github.com/paritytech/polkadot-sdk/assets/10380170/e3c0a7e1-0133-4963-b756-2af4a73a4f5f)

ggwpez commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+//!
+//! Using floating point number types in the runtime should be avoided,
+//! as a single nondeterministic result could cause chaos for consensus along with the
+//! aforementioned issues. For more on the specifics of the peculiarities of floating point calculations, [watch this video by the Computerphile](https://www.youtube.com/watch?v=PZRI1IfStY0)
```
```suggestion
//! aforementioned issues. For more on the specifics of the peculiarities of floating point calculations, [watch this video by the Computerphile](https://www.youtube.com/watch?v=PZRI1IfStY0).
```

juangirini commented:
the base of the PR has changed, it shouldn't create/update the `developer-hub` dir, but rather `docs/sdk`

ggwpez commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+		// Rust includes `.get`, returning Option<T> - so lets use that:
+		assert_eq!(my_list.get(5), None)
+	}
+	#[docify::export]
```
```suggestion

	#[docify::export]
```

ggwpez commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+
+	#[docify::export]
+	#[test]
+	fn bad_collection_retrieval() {
```
There should probably be a `should_panic` here?

ggwpez commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+	}
+	#[docify::export]
+	#[test]
+	#[should_panic(expected = "Defensive failure has been triggered!")]
```
```suggestion
	#[cfg_attr(debug_assertions, should_panic(expected = "Defensive failure has been triggered!")]
```
As you said above: it only panics with debug assertions and would fail on `cargo test --release`.

gpestana commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+//!
+//! Defensive programming is a design paradigm that enables a particular program to continue
+//! running despite unexpected behavior. These unforseen circumstances may
+//! cause the program to stop, (or in the Rust context, `panic!`) defensive practices allow for
```
```suggestion
//! cause the program to stop or, in the Rust context, `panic!`. Defensive practices allow for
```

gpestana commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+//! Defensive programming is a design paradigm that enables a particular program to continue
+//! running despite unexpected behavior. These unforseen circumstances may
+//! cause the program to stop, (or in the Rust context, `panic!`) defensive practices allow for
+//! these circumstances to be accounted for ahead of time, and for them to be handled in a more
```
```suggestion
//! these circumstances to be accounted for ahead of time and for them to be handled in a more
```

nit but afaik the command before "and" is only used when enumerating, right? In any case, either way goes.

gpestana commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+//! these circumstances to be accounted for ahead of time, and for them to be handled in a more
+//! graceful manner.
+//!
+//! The Polkadot SDK is both built to reflect these principles, and to be used accordingly.
```
```suggestion
//! The Polkadot SDK is both built to reflect these principles and to be used accordingly.
```

gpestana commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+//!
+//! When developing within the context of the a runtime, there is one golden rule:
+//!
+//! ***DO NOT PANIC,*** most of the time. There are some exceptions, such as critical operations being actually more
```
```suggestion
//! ***DO NOT PANIC***. There are some exceptions, such as critical operations being actually more
```

Suggesting this change as "do no panic, most of the time" reads funny to me.

gpestana commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+//!
+//! ## General Practices
+//!
+//! When developing within the context of the a runtime, there is one golden rule:
```
Since this ref doc is describing safe programming in Substrate in general, I think it would be interesting to elaborate *why* the runtime should not panic (no fee DoS and, in some cases, stalled runtime) and where it's OK.

CrackTheCode016 replied:
Added one example, it may need more elaboration.

gpestana commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+//! Defensive methods use [`debug_assertions`](https://doc.rust-lang.org/reference/conditional-compilation.html#debug_assertions), which panic in development, but in
+//! production/release, they will merely log an error (i.e., `log::error`).
+//!
+//!
```
```suggestion
```

gpestana commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+    increase_balance_result
+)]
+//!
+//!
```
```suggestion
```

gpestana commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+//! dividing/modulo by zero at any point should be mitigated. One should be, instead, opting for a
+//! `checked_` method in order to introduce safe arithmetic in their code.
+//!
+//!
```
```suggestion
```

gpestana commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+//!
+//! Known scenarios that could be fallible should be avoided: i.e., avoiding the possibility of
+//! dividing/modulo by zero at any point should be mitigated. One should be, instead, opting for a
+//! `checked_` method in order to introduce safe arithmetic in their code.
```
```suggestion
//! `checked_*` method in order to introduce safe arithmetic in their code.
```

gpestana commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+//! consider some other ways of dealing with runtime arithmetic, such as saturated or checked
+//! operations, that won't carry these potential consequences.
+//!
+//!
```
```suggestion
```

gpestana commented on developer-hub/src/reference_docs/defensive_programming.rs:
````diff
@@ -0,0 +1,558 @@
+//! ```
+//!
+//! </details>
+//!
````
```suggestion
```

gpestana commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+//! </details>
+//!
+//!
+//! From the above, we can clearly see the problematic nature of seemingly simple operations in
```
```suggestion
//! From the above, we can clearly see the problematic nature of seemingly simple operations in the
```

gpestana commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+//! [`PerThing`](sp_arithmetic::PerThing) are used:
+//! - **[`Perbill`](sp_arithmetic::Perbill), parts of a billion**
+#![doc = docify::embed!("./src/reference_docs/defensive_programming.rs", perbill_example)]
+//!
```
```suggestion
```

gpestana commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+//! percentage of a particular item:
+#![doc = docify::embed!("./src/reference_docs/defensive_programming.rs", percent_mult)]
+//!
+//!
```
```suggestion
```

kianenigma commented on developer-hub/src/reference_docs/mod.rs:
```diff
@@ -49,7 +49,7 @@ pub mod frame_origin;
 
 /// Learn about how to write safe and defensive code in your FRAME runtime.
 // TODO: @CrackTheCode016 https://github.com/paritytech/polkadot-sdk-docs/issues/44
```
```suggestion
```

kianenigma commented on developer-hub/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,558 @@
+//! running despite unexpected behavior. These unforseen circumstances may
+//! cause the program to stop, (or in the Rust context, `panic!`) defensive practices allow for
+//! these circumstances to be accounted for ahead of time, and for them to be handled in a more
+//! graceful manner.
```
Link to the wikipedia page of the concept would be good.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4865570

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4866295

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4897867

franciscoaguirre commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,589 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! As our runtime should _never_ panic; this includes carefully handling [`Result`]/[`Option`]
```
```suggestion
//! As our runtime should _never_ panic, we should carefully handle [`Result`]/[`Option`]
```

I'd change the person on this paragraph. Sounds better that way.

franciscoaguirre commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,589 @@
+//! even replacing floating point usage with fixed point arithmetic to mitigate issues that come
+//! with floating point calculations.
+//!
+//! Intentional and predictable design should be our first and foremost
+//! property for ensuring a well running, safely designed system.
```
```suggestion
//! Intentional and predictable design should be our first and foremost
//! priority for ensuring a well running, safely designed system.
```

The property of predictability should be our priority.

franciscoaguirre commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,589 @@
+//! ## Defensive Programming
+//!
+//! [Defensive programming](https://en.wikipedia.org/wiki/Defensive_programming) is a design paradigm that enables a particular program to continue
+//! running despite unexpected behavior. These unforseen circumstances may
```
```suggestion
//! running despite unexpected behavior. Unforseen circumstances may
```

franciscoaguirre commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,589 @@
+//! these circumstances to be accounted for ahead of time and for them to be handled in a more
+//! graceful manner.
+//!
+//! The Polkadot SDK is both built to reflect these principles and to be used accordingly.
```
```suggestion
//! The Polkadot SDK is both built to reflect these principles and to facilitate their usage.
```

franciscoaguirre commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,589 @@
+//!
+//!  General guidelines:
+//!
+//! - Avoid any writing functions that could explicitly panic.  Directly using `unwrap()` for a
```
```suggestion
//! - Avoid writing functions that could explicitly panic.  Directly using `unwrap()` for a
```

franciscoaguirre commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,589 @@
+//!  General guidelines:
+//!
+//! - Avoid any writing functions that could explicitly panic.  Directly using `unwrap()` for a
+//!   [`Result`] or common errors such as accessing an out of bounds index on a collection (using
+//!   safer methods to access collection types, i.e., `get()`) are both examples of what to
```
I'd move the recommendation inside parenthesis to a phrase after listing both examples.

franciscoaguirre commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,589 @@
+//!   performed a check beforehand) that a value won't panic upon unwrapping.  Even this is
+//!   discouraged, however, as future changes to that function could then cause that statement to
+//!   panic.  It is better to ensure all errors are propagated and handled accordingly in some way.
+//! - If a function *can* panic, it usually is prefaced with `unchecked_` to indicate its safety.
```
```suggestion
//! - If a function *can* panic, it usually is prefaced with `unchecked_` to indicate its unsafety.
```

franciscoaguirre commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,589 @@
+//!
+//! <details>
+//!   <summary><b>Solution: Saturating or Checked</b></summary>
+//!     For Bob's balance problems, using a `saturated_add` or `checked_add` could've mitigated this
```
```suggestion
//!     For Bob's balance problems, using a `saturating_add` or `checked_add` could've mitigated this
```

franciscoaguirre commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,589 @@
+//! Each of these can be used to construct and represent ratios within our runtime.
+//! You will find types like [`Perbill`](sp_arithmetic::Perbill) being used often in pallet
+//! development.  [`pallet_referenda`] is a good example of a pallet which makes good use of fixed
+//! point arithmetic to calculate.
```
... to calculate what? The sentence makes me think something should come after this

franciscoaguirre commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,589 @@
+//! point arithmetic to calculate.
+//!
+//! Let's examine the usage of `Perbill` and how exactly we can use it as an alternative to floating
+//! point numbers in Substrate development. For this scenario, let's say we are demonstrating a
```
```suggestion
//! point numbers in development with Substrate. For this scenario, let's say we are demonstrating a
```

franciscoaguirre commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,589 @@
+//! _voting_ system which depends on reaching a certain threshold, or percentage, before it can be
+//! deemed valid.
+//!
+//! For most cases, `Perbill` gives us a reasonable amount of precision for most applications, which
```
```suggestion
//! For most applications, `Perbill` gives us a reasonable amount of precision, which
```

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4906495

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4906634

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4923307

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//! running despite unexpected behavior, input or events which may arise in runtime. Normally,
+//! unforeseen circumstances may cause the program to stop or, in the Rust context, `panic!`.
+//! Defensive practices allow for these circumstances to be accounted for ahead of time and for them
+//! to be handled in a graceful manner, which is in the line of the intended, fault-tolerant
+//! behavior of blockchains.
```
```suggestion
//! to be handled in a graceful manner, which is in the line of the intended, fault-tolerant and deterministic
//! behavior of blockchains.
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//! to be handled in a graceful manner, which is in the line of the intended, fault-tolerant
+//! behavior of blockchains.
+//!
+//! The Polkadot SDK is both built to reflect these principles and to facilitate their usage
+//! accordingly.
```
```suggestion
//! The Polkadot SDK is built to reflect these principles and to facilitate their usage
//! accordingly.
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//!
+//! When developing within the context of the a runtime, there is *one* golden rule:
+//!
+//! ***DO NOT PANIC***. There are some exceptions, such as critical operations being actually more
+//! dangerous than allowing the runtime to continue functioning (block authoring, consensus, etc).
```
Please rephrase this sentence. Unable to understand what the exception is here - do we want to panic sometimes deliberately?

EDIT: I see that this is explained in detail later on. Better to mention that and delete the content here.

```suggestion
//! ***DO NOT PANIC***. There are some exceptions, which will be covered later on in this doc.
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//! > It's important to make the differentiation between the **runtime** and **node**.  The runtime
+//! > refers to the core business logic of a Substrate-based chain, whereas the node refers to the
+//! > outer client which deals with telemetry and gossip from other nodes. For more information,
+//! > read about Substrate's architecture.
```
```suggestion
//! > read about Substrate node's architecture.
```

kianenigma replied:
There should be a link here to the other ref doc page explaining runtime vs node.

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//! > refers to the core business logic of a Substrate-based chain, whereas the node refers to the
+//! > outer client which deals with telemetry and gossip from other nodes. For more information,
+//! > read about Substrate's architecture.
+//! > It's also important to note that the behavior of the **node** may differ from that of the
+//! > **runtime**, which is also why at times, you may see `unwrap()` or other "non-defensive"
+//! > behavior taking place.
```
```suggestion
//! > It's also important to note that the criticality of the **node** is slightly lesser than that of the
//! > **runtime**, which is why in a few places of the node's code repository, you may see `unwrap()` or other "non-defensive"
//! > code instances.
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//!
+//!  General guidelines:
+//!
+//! - Avoid writing functions that could explicitly panic. Directly using `unwrap()` for a
+//!   [`Result`], or common errors such as accessing an out of bounds index on a collection should
+//!   not be used. Safer methods to access collection types, i.e., `get()` are available, upon which
+//!   defensive handling of the resulting [`Option`] can occur.
```
```suggestion
//! - Avoid writing functions that could explicitly panic. Directly using `unwrap()` on a
//!   [`Result`], or  accessing an out-of-bounds index on a collection, should
//!   be avoided. Safer methods to access collection types, i.e., `get()` which allow
//!   defensive handling of the resulting [`Option`] are recommended to be used.
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//! - It may be acceptable to use `except()`, but only if one is completely certain (and has
+//!   performed a check beforehand) that a value won't panic upon unwrapping.  Even this is
+//!   discouraged, however, as future changes to that function could then cause that statement to
+//!   panic.  It is better to ensure all errors are propagated and handled accordingly in some way.
```
```suggestion
//!   panic.  It is important to ensure all possible errors are propagated and handled effectively.
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//! - If you are writing a function that could panic, [be sure to document it!](https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html#documenting-components)
+//! - Carefully handle mathematical operations.  Many seemingly, simplistic operations, such as
+//!   **arithmetic** in the runtime, could present a number of issues [(see more later in this
+//!   document)](#integer-overflow).
```
```suggestion
//!   document)](#integer-overflow). Use checked arithmetic wherever possible.
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
````diff
@@ -0,0 +1,618 @@
+//! }
+//! ```
+//!
+//! There are other such examples throughout various pallets, mostly those who are crucial to the
+//! blockchain's function.
````
```suggestion
//! There are other such examples in various pallets, mostly those that are crucial to the
//! blockchain's functionality.
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//!
+//! ### Defensive Traits
+//!
+//! The [`Defensive`](frame::traits::Defensive) trait provides a number of functions, all of which
```
Perhaps, point to their example usage in the repo?

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//! (from the `MAX` back to zero).
+//!
+//! In the context of runtime development, we don't always have control over what is being supplied
+//! as a parameter. For example, even this simple adding function could present one of two outcomes
```
```suggestion
//! as a parameter. For example, even this simple add function could present one of two outcomes
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//! made obvious, especially in the context of blockchain development, where unsafe arithmetic could
+//! produce unexpected consequences like a user balance over or underflowing.
+//!
+//! Luckily, there are ways to both represent and handle these scenarios depending on our specific
```
```suggestion
//! Fortunately, there are ways to both represent and handle these scenarios depending on our specific
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//! potential for inaccuracy, rounding errors, or other unexpected behavior.
+//!
+//! Using floating point number types in the runtime should be avoided,
+//! as a single nondeterministic result could cause chaos for consensus along with the
```
```suggestion
//! as a single non-deterministic result could cause chaos for blockchain consensus along with the
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//! as a single nondeterministic result could cause chaos for consensus along with the
+//! aforementioned issues. For more on the specifics of the peculiarities of floating point calculations, [watch this video by the Computerphile](https://www.youtube.com/watch?v=PZRI1IfStY0).
+//!
+//! The following methods demonstrate different ways one can handle numbers safely natively in Rust,
+//! without fear of panic or unexpected behavior from wrapping.
```
```suggestion
//! The following methods demonstrate different ways one can handle numbers natively in Rust in a safe manner,
//! without fear of panic or unexpected behavior from wrapping.
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//!
+//! ### Checked Arithmetic
+//!
+//! **Checked operations** utilize an `Option<T>` as a return type. This allows for simple pattern
+//! matching to catch any unexpected behavior in the event of an overflow.
```
```suggestion
//! **Checked operations** utilize an `Option<T>` as a return type. This allows for
//! catching any unexpected behavior in the event of an overflow through simple pattern matching.
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//! This is an example of a valid operation:
+#![doc = docify::embed!("./src/reference_docs/defensive_programming.rs", checked_add_example)]
+//!
+//! This is an example of an invalid operation, in this case, a simulated integer overflow, which
+//! would simply result in `None`:
```
```suggestion
//! This is an example of an invalid operation. In this case, a simulated integer overflow, which
//! would simply result in `None`:
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//! operations are a safe bet, as it presents two, predictable (and _erroring_) outcomes that can be
+//! handled accordingly (`Some` and `None`).
+//!
+//! In a practical context, the resulting [`Option`] should be handled accordingly. The following
+//! conventions can be seen from the within the Polkadot SDK, where it can be handled in one of
+//! two ways:
```
```suggestion
//! In a practical context, the resulting [`Option`] should be handled accordingly. The following
//! conventions can be seen within the Polkadot SDK, where it is handled in
//! two ways:
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//!
+//! #### Handling via Result - Less Verbose
+//!
+//! In the Polkadot SDK codebase, you may see checked operations being handled as a [`Result`] via
+//! `ok_or`. This is a less verbose way of expressing the above.  Which to use often boils down to
+//! the developer's preference:
```
```suggestion
//! In the Polkadot SDK codebase, you may see checked operations being handled as a [`Result`] via
//! `ok_or`. This is a less verbose way of expressing the above. This usage often boils down to
//! the developer's preference:
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//!
+//! ### Saturating Operations
+//!
+//! Saturating a number limits it to the type's upper or lower bound, no matter the integer would
+//! overflow in runtime. For example, adding to `u32::MAX` would simply limit itself to `u32::MAX`:
```
```suggestion
//! Saturating a number limits it to the type's upper or lower bound, even if the integer were to
//! overflow in runtime. For example, adding to `u32::MAX` would simply limit itself to `u32::MAX`:
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//!
+//!
+//! Known scenarios that could be fallible should be avoided: i.e., avoiding the possibility of
+//! dividing/modulo by zero at any point should be mitigated. One should be, instead, opting for a
+//! `checked_*` method in order to introduce safe arithmetic in their code.
```
```suggestion
//! dividing/modulo by zero at any point should be mitigated. One should be opting for a
//! `checked_*` method to introduce safe arithmetic in their code.
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//! #### The problem with 'default' wrapped operations
+//!
+//! **Wrapped operations** cause the overflow to wrap around to either the maximum or minimum of
+//! that type. Imagine this in the context of a blockchain, where there are balances, voting
```
```suggestion
//! that type. Imagine this in the context of a blockchain, where there are account balances, voting
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//! that type. Imagine this in the context of a blockchain, where there are balances, voting
+//! counters, nonces for transactions, and other aspects of a blockchain.
+//!
+//! Some of these mechanisms can be more critical than others. It's for this reason that we may
+//! consider some other ways of dealing with runtime arithmetic, such as saturated or checked
+//! operations, that won't carry these potential consequences.
+//!
+//! While it may seem trivial, choosing how to handle numbers is quite important. As a thought
+//! exercise, here are some scenarios of which will shed more light on when to use which.
```
Removing redundant info that was already presented and emphasized before.

```suggestion
//!
//! While it may seem trivial, choosing how to handle numbers is quite important. As a thought
//! exercise, here are some scenarios of which will shed more light on when to use which.
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//!   as the notion of balances, are often shared across multiple pallets - meaning these small
+//!   changes can make a big difference in outcome. </details>
+//!
+//! #### Proposals' ID Overwrite
```
```suggestion
//! #### Proposal ID Overwrite
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//! The type for counting the number of proposals on-chain is represented by a `u8` number, called
+//! `proposals_count`. Every time a new proposal is added to the system, this number increases. With
+//! the proposal pallet being high in usage, it has reached `u8::MAX`'s limit of `255`, causing
+//! `proposals_count` to go to `0`. Unfortunately, this resulted in new proposals overwriting old
```
```suggestion
//! `proposals_count` to go to `0`. Unfortunately, this results in new proposals overwriting old
```

DrW3RK commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//! </details>
+//!
+//! From the above, we can clearly see the problematic nature of seemingly simple operations in the
+//! runtime. Of course, it may be that using unchecked math is perfectly fine under some scenarios -
+//! such as certain balance being never realistically attainable, or a number type being so large
+//! that it could never realistically overflow unless one sent thousands of transactions to the
+//! network.
```
Neither of the presented scenarios warrant usage of unchecked math. If there are situtations that limit someone from using checked math, they should be highlighted here. Else, we don't say anything.

```suggestion
//! runtime. 
```

DrW3RK reviewed (commented):
Improving readability.

kianenigma commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//! ***DO NOT PANIC***. There are some exceptions, such as critical operations being actually more
+//! dangerous than allowing the runtime to continue functioning (block authoring, consensus, etc).
+//!
+//! > It's important to make the differentiation between the **runtime** and **node**.  The runtime
```
You can reference to the `substrate_architecture`, `wasm_metaprotocol` or similar pages that describe this, if you prefer brevity.

kianenigma commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//!
+//!  General guidelines:
+//!
+//! - Avoid writing functions that could explicitly panic. Directly using `unwrap()` for a
```
You can showcase an example of this.

```rust
fn bad_pop<T>(v: Vec<T>) -> T {} 
fn good_pop<T>(v: Vec<T>) -> Option<T> {}
```

kianenigma commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//!   **arithmetic** in the runtime, could present a number of issues [(see more later in this
+//!   document)](#integer-overflow).
+//!
+//! ### Examples of when to `panic!`
```
I would hesitate mentioning this toward the beginning of this article, as it could be misleading. 

95% of pallet developers are not going to write pallets that are consensus level, and are therefore never in need of having an explicit panic. So I would first address the majority, and leave this as some kind of appendix.

CrackTheCode016 replied:
I placed it at the end of the doc and specified this point.

kianenigma commented on docs/sdk/src/reference_docs/defensive_programming.rs:
```diff
@@ -0,0 +1,618 @@
+//! and handle values.  This can be used in place of
+//! an `expect`, and again, only if the developer is sure about the unwrap in the first place.
+//!
+//! Here is a full list of all defensive types:
```
We have a `mod defensive` in frame-support (and if not, we can create one), so I would not repeat this list here and instead link to it.

kianenigma reviewed (approved):
Thanks for your PR. I have a few suggestions: 

* There are two topics being discussed here:
  1. How to deal with panics, a subset of which is also how to deal with `Option` and `Result`
  2. How to deal with `Fixed` and `PerThing`, which IMO does not have much to do with the former and defensive programming. 

To make the page more concise, I suggest removing the latter as it is a bit off-topic. I think all of your work and examples on these types are valuable and should stay, but should merely be added as more information and examples to the rust-doc of `per_thing.rs`, and at most linked from here. 

* The same can be said about the rest of the snippets of the code as well. We are (re)listing functions and types in rust modules here, instead of linking. Did you consider changing the source code files of `defensive.rs` and `sp-arithmetic` while doing this, or intentionally prevented yourself from doing so? It seems to me that if you had split parts of your changes in the code rust files and this doc, the outcome would have been better. 

With that said, I can summarize my comment as: "The content is good, but it is a) a bit unorganized b) a bit more verbose than what it could be".

In general though, I would approve it already. I do expect you to resolve the structure of the page before merging though, as in it should cover only one topic, and not two.

TLDR; should try and be more concise, and contribute more to code files, and less to the doc files. But it is a good start.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5204391

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5204408

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5204409

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5204410

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5204425

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: check-tracing
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5204426

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5204809

kianenigma commented:
Should be ready for merge once CI is green and an approval from @gpestana.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5230677

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5328904

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5478521

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5495307

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5495453

CrackTheCode016 commented:
@kianenigma I have hopefully addressed your feedback, added documentation to the supporting files in `sp_arithmetic`, and severely reduced the size of the reference doc without sacrificing the content.  It may still be verbose in some areas; let me know if this is good for merging. Thanks!

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5495744

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5522011

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5558037

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5574501

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5574536

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5575219

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5575367

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5575670

CrackTheCode016 commented:
@kianenigma (sorry for the ping) CI looks all good and green now.

kianenigma commented:
> @kianenigma (sorry for the ping) CI looks all good and green now.

You need to fix the toml formats it seems.
